pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentStickyKeyDispatcherMultipleConsumers.java

```diff
@@ -0,0 +1,159 @@
+                    log.info("[{}] rewind because no available consumer found for key {} from total {}", name,
+                            entriesWithSameKey.getKey(), consumerList.size());
+                    entriesWithSameKey.getValue().forEach(Entry::release);
+                    cursor.rewind();
+                    return;
+                }
+
+                int messagesForC = Math.min(entriesWithSameKey.getValue().size(), consumer.getAvailablePermits());
+                if (messagesForC > 0) {
+
+                    // remove positions first from replay list first : sendMessages recycles entries
+                    List<Entry> subList = new ArrayList<>(entriesWithSameKey.getValue().subList(0, messagesForC));
+                    if (readType == ReadType.Replay) {
+                        subList.forEach(entry -> messagesToReplay.remove(entry.getLedgerId(), entry.getEntryId()));
+                    }
+                    final SendMessageInfo sentMsgInfo = consumer.sendMessages(subList);
+                    entriesWithSameKey.getValue().removeAll(subList);
+                    final long msgSent = sentMsgInfo.getTotalSentMessages();
+                    totalAvailablePermits -= msgSent;
+                    totalMessagesSent += sentMsgInfo.getTotalSentMessages();
+                    totalBytesSent += sentMsgInfo.getTotalSentMessageBytes();
+
+                    if (entriesWithSameKey.getValue().size() == 0) {
+                        iterator.remove();
+                    }
+                }
+            }
+
+            // acquire message-dispatch permits for already delivered messages
+            if (serviceConfig.isDispatchThrottlingOnNonBacklogConsumerEnabled() || !cursor.isActive()) {
+                if (topic.getDispatchRateLimiter().isPresent()) {
+                    topic.getDispatchRateLimiter().get().tryDispatchPermit(totalMessagesSent, totalBytesSent);
+                }
+
+                if (dispatchRateLimiter.isPresent()) {
+                    dispatchRateLimiter.get().tryDispatchPermit(totalMessagesSent, totalBytesSent);
+                }
+            }
+
+            if (groupedEntries.size() > 0) {
+                int laterReplay = 0;
+                for (List<Entry> entryList : groupedEntries.values()) {
+                    laterReplay += entryList.size();
+                    entryList.forEach(entry -> {
+                        messagesToReplay.add(entry.getLedgerId(), entry.getEntryId());
+                        entry.release();
+                    });
+                }
+                if (log.isDebugEnabled()) {
+                    log.debug("[{}] No consumers found with available permits, storing {} positions for later replay", name,
+                            laterReplay);
+                }
+
+            }
+        }
+    }
+
+    @Override
+    public SubType getType() {
+        return SubType.Key_Shared;
+    }
+
+    private String peekStickyKey(ByteBuf metadataAndPayload) {
```

Comment:
this doesn't seem to handle batched messages. any thoughts on how we can handle that?

Reply:
I think we don't need to handle batched messages here, for now batched message without key, so this will be treated as `NONE_KEY` and all message without key will be send to a consumer, i will make a PR for `NONE_KEY` policy setting by client.

For batched messages, we need a key based message batching mechanism, by this way peekStickyKey() still using the same logic.

`NONE_KEY` policy and key based message batching mechanism is already in my backlog [ISSUE-4077](https://github.com/apache/pulsar/issues/4077) .